### PR TITLE
Add a Maybe validator

### PR DESCRIPTION
**Pull request by AndreaCrotti**
Simple Maybe validator, which encapsulates the fact that something can be None or of the given type

### CODE REVIEW

**Comment by AndreaCrotti** [comment]
It's quite a standard constructor in functional languages, I think it might be useful here as well since it's a common case for functions to return None in some conditions..

File: voluptuous/tests/tests.py [hunk 1]
```diff
@@ -544,6 +544,19 @@ def fn(arg):
     fn(1)
 
 
+def test_schema_decorator_no_args():
+    @validate(int, __return__=Maybe(int))
```
Comment (tusharmakkar08) [inline voluptuous/tests/tests.py]: Can you add test case for Invalid input as well?

File: voluptuous/validators.py [hunk 2]
```diff
@@ -459,6 +459,14 @@ def PathExists(v):
         raise PathInvalid("Not a Path")
 
 
+class Maybe(object):
```
Comment (tusharmakkar08) [inline voluptuous/validators.py]: Add `__repr__` and docstring.

**Comment by AndreaCrotti** [comment]
Yes sure @tusharmakkar08 I will solve the issues, just wanted to know if you were happy to add this at all before I made the PR fully finished...
Is that something you would be happy to add then?

**Comment by tusharmakkar08** [comment]
Yeah, `Maybe` sounds like a decent validator which incorporates `None` or the type which is specified in the schema. Although it can be done using `Any` but I believe `Maybe` addition would make it more verbose and user-friendly.

**Comment by alecthomas** [comment]
Is this any different from Any(T, None)?

If not, I'm -1 on this as it adds only very minor convenience.

**Comment by tusharmakkar08** [comment]
AFAIK it's same as Any(T, None) just more verbose. @AndreaCrotti : Please confirm this.

**Comment by AndreaCrotti** [comment]
Well in terms of the result I guess it's the same thing, but:

1. Maybe is a constructor that everyone coming from functional languages will understand and be used to.
2. Maybe is in theory much more restrictive, and in because of this more clear what it actually does.
    Since in theory if you pass a type it just means that the value can be of the given type or None, nothing else, while with Any you can pass any callable.

We can write Maybe using Any if you prefer, but I generally think it would be a nice thing to add..

**Comment by AndreaCrotti** [comment]
Any news about this @alecthomas ?

I'm happy to implement it using Any internally if you prefer and add the missing tests/documentation..

**Comment by alecthomas** [comment]
Sure, why not. Pretty minor change but if it helps with clarity it is probably a good idea.

**Comment by AndreaCrotti** [comment]
@tusharmakkar08 @alecthomas I've done a few changes, can you let me know what you think now?

strangely I wanted to add a test like this:

        assert_equal(repr(s), "Maybe(<class 'int'>)")

but for some reason it doesn't work, even if I get exactly that  trying from ipython..

**Comment by AndreaCrotti** [comment]
Also I've added a check to make sure that what you can only pass an actual type.

I could actually use the @validate decorator if you are happy about that.

**Comment by tusharmakkar08** [comment]
It's surprising that `__repr__` test isn't working. I will try to look into it maybe later today. Feel free to use `@validate` decorator there.

**Comment by AndreaCrotti** [comment]
Very strangely if I try to do this

```
class Maybe(object):
    @validate(arg=type)
    def __init__(self, kind, msg=None):
        self.kind = kind

```

I get a very strange error 
```
----------------------------------------------------------------------
File "/home/andrea/projects/oss_projects/python/voluptuous/README.md", line 568, in README.md
Failed example:
    str(exc) == "not a valid value @ data[0][0]"
Expected:
    True
Got:
    False
----------------------------------------------------------------------
File "/home/andrea/projects/oss_projects/python/voluptuous/README.md", line 578, in README.md
Failed example:
    schema([6])
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest README.md[106]>", line 1, in <module>
        schema([6])
    NameError: name 'schema' is not defined


----------------------------------------------------------------------
```

**Comment by AndreaCrotti** [comment]
I can also leave it like this otherwise if you are happy with it, and add a bit of documentation in case it's all good to go

**Comment by tusharmakkar08** [comment]
Hey @AndreaCrotti 

I would ideally like to have a test case with the `__repr__` test. Once you get that working, please squash your commits and we would be good to go. 

Thanks.

**Comment by AndreaCrotti** [comment]
I've done the repr test but it's a bit trivial now, because to support both py2 and py3 and fix this error
```
FAIL: Verify that __repr__ returns valid Python expressions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.2.6/lib/python3.2/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/alecthomas/voluptuous/voluptuous/tests/tests.py", line 384, in test_repr
    assert_equal(repr(maybe_int), "Maybe(<type 'int'>)")
AssertionError: "Maybe(<class 'int'>)" != "Maybe(<type 'int'>)"
- Maybe(<class 'int'>)
?        ^^^^^
+ Maybe(<type 'int'>)
?        ^^^^
```

I had to just repr(int) inside the test itself

**Comment by AndreaCrotti** [comment]
@tusharmakkar08 @alecthomas what do you think better now?

**Comment by tusharmakkar08** [comment]
@AndreaCrotti I would prefer checking which python version is there and then equating the `__repr__`. Something similar has been done by @jmwri [here](https://github.com/alecthomas/voluptuous/blob/master/voluptuous/util.py#L162).

**Comment by AndreaCrotti** [comment]
Uhm yes ok I can do that sure, but any reason in general to not use the `six` library as most of other projects do?
I mean in my particular case I'm not sure it would help that much (at least I didn't see it) but most cases it's exactly what you need to write 2/3 compatible code..

**Comment by tusharmakkar08** [comment]
Yeah, We do not want to have any external dependency within the code AFAP.

**Comment by AndreaCrotti** [comment]
Yes ok that's fine @tusharmakkar08 
however in this specific case the line I wrote is
```
assert_equal(repr(maybe_int), "Maybe(%s)" % str(int))
```
I can potentially as you say do some python version dependent conditional to return one or the other, but it would mean I'm testing what Python2/3 return with str(int), which is kind of pointless imho..
This test still ensures we get the right result, even if a bit trivial, but the alternative doesn't make much more sense to me to be honest.

**Comment by tusharmakkar08** [comment]
No, you don't have to do that. You can just check for `sys.version_info` and then update the suitable test cases.

**Comment by AndreaCrotti** [comment]
you mean by doing something like this @tusharmakkar08 ?

```
if sys.version_info[0] == 2:
   assert_equal(repr(maybe_int), "Maybe('type<int>')"
elif sys.version_info[0] == 3:
   assert_equal(repr(maybe_int), "Maybe('class<int>')"

```

I can do it of course but just wantedto clarify that it's rather pointless, since there it's just also testing what type(int) does on Python2/Python3, which is not really something we need to test imho..

**Comment by tusharmakkar08** [comment]
@AndreaCrotti Please refer [here](https://github.com/alecthomas/voluptuous/blob/master/voluptuous/util.py#L162) and update your PR correspondingly.

**Comment by AndreaCrotti** [comment]
@tusharmakkar08 that's not related at all I believe..
I'm not changing behaviour depending on the python version, just the actual output from the test changes, so that doesn't really apply..

File: voluptuous/validators.py [hunk 3]
```diff
@@ -459,6 +459,35 @@ def PathExists(v):
...
+    """Validate that the object is of a given type or is None.
+
+    :raises Invalid: if the value is not of the type declared or None
```
Comment (divanovGH) [inline voluptuous/validators.py]: This contradicts what is written above. None type should not raise an exception.

**Comment by AndreaCrotti** [comment]
Just addressed the comments from @divanovGH 
Anything else I should do for this @tusharmakkar08 @alecthomas ?

**Comment by tusharmakkar08** [comment]
Hey @AndreaCrotti 

Please squash the commits. 

Thanks

**Pull request merged**